docs/proposals/bulward-integration.md

```diff
@@ -0,0 +1,206 @@
+A possible workaround would be using two `Account`s for `couchdb-consumer` and `redisdb-operator`, but that just misses the point
+to have the option to make an `Account` be Provider and Tenant at the same time.
+to solve this issue, we can use `Organization` to represent `Account`:
+```yaml
+apiVersion: apiserver.bulward.io/v1alpha1
+kind: Organization
+metadata:
+  name: company-a
+spec:
+  metadata:
+    displayName: The A Company
+    description: In 1972, a crack commando unit was sent to prison by a military court...
+  owners:
+  - kind: User
+    name: company-owner
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User
+    name: couchdb-consumer
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User
+    name: redisdb-operator
+    apiGroup: rbac.authorization.k8s.io
+```
+And then we can have dedicated `Project` for both `couchdb` and `redisdb`:
+```yaml
+apiVersion: apiserver.bulward.io/v1alpha1
+kind: Project
+metadata:
+  name: couchdb
+  namespace: organization-a
+spec:
+  owners:
+  - kind: User
+    name: owner
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User
+    name: couchdb-consumer
+    apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: apiserver.bulward.io/v1alpha1
+kind: Project
+metadata:
+  name: redisdb
+  namespace: organization-a
+spec:
+  owners:
+  - kind: User
+    name: owner
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User
+    name: redisdb-operator
+    apiGroup: rbac.authorization.k8s.io
+```
+In this way, we can easily grant different permissions for different team, and different projects. For example, in bulward,
+we can use `ProjectRoleTemplate` and `ProjectRole` to set up permission for a (group of) `Project`.
+
+Overall:
+- `Account` should be mapped to `Organzation` in a way, not `Project`.
+- `Organization` can be `Provider` and `Tenant` at the same time, but `Project` under
+`Organization` should be defined as Provider project or Tenant project.
+
+## Implementation
+### Deployment
+Every bulward release can contain a manifest yaml file, just like cert-manager, we can install it in the similar way:
+```bash
+kubectl apply -f https://github.com/kubermatic/bulward/releases/download/${BULWARD_VERSION}/bulward.yaml
+kubectl wait --for=condition=available deployment/bulward-controller-manager -n bulward-system --timeout=120s
+kubectl wait --for=condition=available deployment/bulward-apiserver -n bulward-system --timeout=120s
+```
+### API Server
+#### Organization Endpoint
+We have an endpoint for `Account`, we can reuse it for `Organization`, also, in Bulward, only `Organization` object is visible
+for organization member (this is implemented in Bulward api extension server), we can also drop the filtering implementation in currently account endpoint.
+#### Project Endpoint
+Project Endpoint (Get/List/Watch/Update/Delete) should be implemented, all permission filtering are offloaded to Bulward api
+extension server, so this will also be a thin layer.
```

Comment:
plus we have to inject the user impersionation in the API server as well.

Reply:
yeah, that's right, will update the proposal

Reply:
done